rasterio/shutil.pyx

```diff
@@ -0,0 +1,206 @@
+
+
+@ensure_env
+def exists(path):
+
+    """Determine if a dataset exists by attempting to open it.
+
+    Parameters
+    ----------
+    path : str
+        Path to dataset.
+    """
+
+    cdef GDALDatasetH h_dataset = NULL
+
+    gdal_path = vsi_path(*parse_path(path))
+    b_path = gdal_path.encode('utf-8')
+    cdef char* c_path = b_path
+
+    with nogil:
+        h_dataset = GDALOpenShared(c_path, <GDALAccess>0)
+
+    try:
+        h_dataset = exc_wrap_pointer(h_dataset)
+        return True
+    except CPLE_OpenFailedError:
+        return False
+    finally:
+        GDALClose(h_dataset)
+
+
+@ensure_env
+def copy(src, dst, driver='GTiff', strict=True, **creation_options):
+
+    """Copy a raster from a path or open dataset handle to a new destination
+    with driver specific creation options.
+
+    Parameters
+    ----------
+    src : str or rasterio.io.DatasetReader
+        Path to source dataset or open dataset handle.
+    dst : str
+        Output dataset path.
+    driver : str, optional
+        Output driver name.
+    strict : bool, optional.  Default: True
+        Indicates if the output must be strictly equivalent or if the
+        driver may adapt as necessary.
+    creation_options : **kwargs, optional
+        Creation options for output dataset.
+    """
+
+    cdef bint c_strictness
+    cdef char **options = NULL
+    cdef char* c_src_path = NULL
+    cdef char* c_dst_path = NULL
+    cdef GDALDatasetH src_dataset = NULL
+    cdef GDALDatasetH dst_dataset = NULL
+    cdef GDALDriverH drv = NULL
+
+    # Creation options
+    for key, val in creation_options.items():
+        kb, vb = (x.upper().encode('utf-8') for x in (key, str(val)))
+        options = CSLSetNameValue(
+            options, <const char *>kb, <const char *>vb)
+        log.debug("Option %r:%r", kb, vb)
+
+    c_strictness = strict
+    driverb = driver.encode('utf-8')
+    drv = GDALGetDriverByName(driverb)
+    if drv == NULL:
+        raise DriverRegistrationError("Unrecognized driver: {}".format(driver))
+
+    # Input is a path or GDAL connection string
+    if isinstance(src, str):
+        src = src.encode('utf-8')
+        c_src_path = src
+        with nogil:
+            src_dataset = GDALOpenShared(c_src_path, <GDALAccess>0)
+        src_dataset = exc_wrap_pointer(src_dataset)
+        close_src = True
+    # Input is something like 'rasterio.open()'
+    else:
+        src_dataset = (<DatasetReaderBase?>src).handle()
+        close_src = False
+
+    dst = dst.encode('utf-8')
+    c_dst_path = dst
+    try:
+        with nogil:
+            dst_dataset = GDALCreateCopy(
+                drv, c_dst_path, src_dataset, c_strictness, options, NULL, NULL)
+        dst_dataset = exc_wrap_pointer(dst_dataset)
+    finally:
+        CSLDestroy(options)
+        if close_src:
+            GDALClose(src_dataset)
+
+        if dst_dataset != NULL:
+            GDALClose(dst_dataset)
+
+
+@ensure_env
+def copyfiles(src, dst):
```

Comment:
`$ rio mask` was doing an `os.unlink()`, which lead to this (and an earlier attempt with `DatasetBase.files`).

Reply:
It occurs to me now that `src` should probably also support `BaseDataset()` subclassed objects.

Reply:
On second look this may be asking for threading traps or other unexpected results.  [`GDALCopyDatsetFiles()`](http://www.gdal.org/classGDALDriver.html#a4be070a452d99dde5196fa8e3ecc2e1a) operates on paths, not an open dataset like [`GDALCreateCopy()`](http://www.gdal.org/classGDALDriver.html#a2c897da2a6e25169cccc49ef48797ce1).